pkg/cloud/cloud.go

```diff
@@ -689,13 +687,34 @@ func (c *cloud) DeleteDisk(ctx context.Context, volumeID string) (bool, error) {
 	return true, nil
 }
 
+// Node likely bad device names cache
+// Remember device names that are already in use on an instance and use them last when attaching volumes
+// This works around device names that are used but do not appear in the mapping from DescribeInstanceStatus
+const cacheForgetDelay = 1 * time.Hour
+
+type cachedNode struct {
+	timer          *time.Timer
+	likelyBadNames map[string]struct{}
+}
+
+var cacheMutex sync.Mutex
+var nodeDeviceCache map[string]cachedNode = map[string]cachedNode{}
```

Comment:
I thought about this, and an in-development version of this PR did do it that way, but I ended up deciding on putting it in `cloud` (at least for the initial PR) for two reasons:

1. Our `DeviceManager` is already a complex nightmare that is difficult to reason about, full of unnecessary indirection. I'd like to refactor (and remove large portions of) it in the future, and putting more code in makes that more difficult.
2. It makes writing tests for `cloud.go` very difficult without resorting to mocking the entire `DeviceManager` (which can't be done simply with gomock, it would have to be a mock that actually functioned and thus duplicated large parts of the `DeviceManager` functionality).

Reply:
After an offline discussion with Connor, I agree that it makes more practical sense to keep this code in `cloud` today (even though we introduce more interweaving of concerns), on the condition that we prioritize redesigning the relationship with `cloud.go` and device allocation asap. 

Once we complete that refactoring work, we can extract some of this PR's code to that separate device allocation file (because today, the tight coupling and interwoven call stacks of `DeviceManager` and `cloud`  negate the encapsulation benefits of `DeviceManager`).